src/Simulation/Simulators/SparseSimulator/Native/quantum_state.hpp

```diff
@@ -0,0 +1,1506 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <random>
+#include <cmath>
+#include <functional>
+#include <algorithm>
+#include <list>
+#include <iostream>
+#include <atomic>
+#include <thread>
+#ifdef _OPENMP
+    #include <omp.h>
+#endif
+
+#include "basic_quantum_state.hpp"
+#include "quantum_hash_map.hpp"
+#include "types.h"
+#include "gates.h"
+
+using namespace std::literals::complex_literals;
+
+namespace Microsoft::Quantum::SPARSESIMULATOR
+{
+
+// power of square root of -1
+inline amplitude iExp(int power)
+{
+
+    int p = ((power % 4) + 8) % 4;
+    switch (p)
```

Comment:
I have an impression that this can be simplified to 
```suggestion
    switch (power % 4)
```

Reply:
`power` can also be negative, so one would still have to take care of the sign.